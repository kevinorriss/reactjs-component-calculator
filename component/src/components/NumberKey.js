import React from 'react'
import PropTypes from 'prop-types'

class NumberKey extends React.Component {
    constructor(props) {
        super(props)
        this.handleKeyDown = this.handleKeyDown.bind(this)
        this.state = {
            number: props.number.toString(),
            onClick: props.onClick
        }
    }

    componentDidMount() {
        window.addEventListener('keydown', this.handleKeyDown)
    }
    
    componentWillUnmount() {
        window.removeEventListener('keydown', this.handleKeyDown)
    }

    handleKeyDown(event) {
        if (event.key === this.state.number) {
            this.state.onClick(this.state.number)
        }
    }

    render() {
        return (
            <div className="button button--number" onClick={() => {this.state.onClick(this.state.number)}}><div>{this.state.number}</div></div>
        )
    }
}

NumberKey.propTypes = {
    onClick: PropTypes.func.isRequired,
    number: PropTypes.oneOf([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]).isRequired
}

export default NumberKey