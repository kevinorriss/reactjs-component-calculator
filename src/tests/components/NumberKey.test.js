import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer';
import NumberKey from '../../../component/src/components/NumberKey'

it('should render NumberKey correctly', () => {
    const component = renderer.create(<NumberKey number={0} onClick={jest.fn()} />).toJSON()
    expect(component).toMatchSnapshot()
})

it('should call onClick when corresponding keyboard key pressed', () => {
    const onCick = jest.fn()
    const map = {}
    window.addEventListener = jest.fn((event, cb) => { map[event] = cb })
    shallow(<NumberKey number={0} onClick={onCick} />)
    map.keydown({ key: '0' })
    expect(onCick).toHaveBeenCalledWith('0')
})

it('should call onClick when clicked', () => {
    const onCick = jest.fn()
    const wrapper = shallow(<NumberKey number={0} onClick={onCick} />)
    wrapper.find('.button--number').simulate('click')
    expect(onCick).toHaveBeenLastCalledWith('0')
})