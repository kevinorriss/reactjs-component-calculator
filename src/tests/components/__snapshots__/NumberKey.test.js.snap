// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render NumberKey correctly 1`] = `
<div
  className="button button--number"
  onClick={[Function]}
>
  <div>
    0
  </div>
</div>
`;
